src/Core/Grid/Position/UpdateHandler/DoctrinePositionUpdateHandler.php

```diff
@@ -0,0 +1,131 @@
+
+    /**
+     * @var string
+     */
+    private $dbPrefix;
+
+    /**
+     * @param Connection $connection
+     * @param string $dbPrefix
+     */
+    public function __construct(
+        Connection $connection,
+        $dbPrefix
+    ) {
+        $this->connection = $connection;
+        $this->dbPrefix = $dbPrefix;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCurrentPositions(PositionDefinitionInterface $positionDefinition, $parentId = null)
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->from($this->dbPrefix . $positionDefinition->getTable(), 't')
+            ->select('t.' . $positionDefinition->getIdField() . ', t.' . $positionDefinition->getPositionField())
+            ->addOrderBy('t.' . $positionDefinition->getPositionField(), 'ASC')
+        ;
+
+        if (null !== $parentId && null !== $positionDefinition->getParentIdField()) {
+            $qb
+                ->andWhere('t.' . $positionDefinition->getParentIdField() . ' = :parentId')
+                ->setParameter('parentId', $parentId)
+            ;
+        }
+
+        $positions = $qb->execute()->fetchAll();
+        $currentPositions = [];
+        foreach ($positions as $position) {
+            $positionId = $position[$positionDefinition->getIdField()];
+            $currentPositions[$positionId] = $position[$positionDefinition->getPositionField()];
+        }
+
+        return $currentPositions;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function updatePositions(PositionDefinitionInterface $positionDefinition, array $newPositions)
+    {
+        try {
+            $this->connection->beginTransaction();
+            $positionIndex = 0;
+            foreach ($newPositions as $rowId => $newPosition) {
+                $qb = $this->connection->createQueryBuilder();
+                $qb
+                    ->update($this->dbPrefix . $positionDefinition->getTable())
+                    ->set($positionDefinition->getPositionField(), ':position')
+                    ->andWhere($positionDefinition->getIdField() . ' = :rowId')
+                    ->setParameter('rowId', $rowId)
+                    ->setParameter('position', $positionIndex)
+                ;
+
+                $statement = $qb->execute();
+                if ($statement instanceof Statement && $statement->errorCode()) {
+                    throw new PositionUpdateException(
+                        'Could not update #%i',
+                        'Admin.Catalog.Notification',
+                        [$rowId]
+                    );
+                }
+                ++$positionIndex;
+            }
+            $this->connection->commit();
+        } catch (ConnectionException $e) {
+            $this->connection->rollBack();
+
+            throw new PositionUpdateException(
+                'Could not update.',
+                'Admin.Catalog.Notification',
+                []
+            );
```

Comment:
How is this exception displayed? If it's not caught, what do you think about sending the previous exception as parent?

It seems you're not folowing the Exception constructor signature to make this change easily, but it could be nice to have it for easier debugging.

Reply:
actually not a good idea because IDE does not detect the exception any more and does not expect it to be thrown